src/parameters.jl

```diff
@@ -48,12 +51,25 @@ function flatten(x::Positive)
 end
 
 """
-    Bounded(value::Real, lower_bound::Real, upper_bound::Real)
+    bounded(value::Real, lower_bound::Real, upper_bound::Real)
 
+Constructs a `Bounded`.
 The `value` of a `Bounded` is a `Real` number that is constrained to be within the interval
 (`lower_bound`, `upper_bound`). This is represented in terms of an `unconstrained_value` and 
-a `transform` that maps any value to reals included by (`lower_bound`, `upper_bound`).
+a `transform` that maps any value to reals in (`lower_bound`, `upper_bound`).
 """
+function bounded(value::Real, lower_bound::Real, upper_bound::Real)
```

Comment:
do we need some @assert on the relationship of value/lower_bound/upper_bound. What would happen if somebody construct things like `bounded(-2.0, -0.1, 2.0)`? (same applies to `positive`)

Reply:
If the specification is inconsistent for `bounded` or `positive` then there would be a DomainError when the inverse of transform function is called right before the `AbstractParameter` is created so I think it's safe.

Reply:
True. Bella has a point though -- it would be nice to have an informative error message. If things are heavily nested it will probably make things significantly easier to debug.

Reply:
Default error message:
```julia
julia> bounded(-0.1, 0.0, 1.0)
ERROR: DomainError with -0.09090909091008265:
log will only return a complex result if called with a complex argument. Try log(Complex(x)).
Stacktrace:
 [1] throw_complex_domainerror(::Symbol, ::Float64) at ./math.jl:33
 [2] log(::Float64) at ./special/log.jl:285
 [3] logit at /Users/willtebbutt/.julia/packages/StatsFuns/CXyCV/src/basicfuns.jl:83 [inlined]
 [4] _logit at /Users/willtebbutt/.julia/packages/Bijectors/66bcj/src/bijectors/logit.jl:24 [inlined]
 [5] Logit at /Users/willtebbutt/.julia/packages/Bijectors/66bcj/src/bijectors/logit.jl:18 [inlined]
 [6] bounded(::Float64, ::Float64, ::Float64) at /Users/willtebbutt/.julia/dev/ParameterHandling/src/parameters.jl:79
 [7] top-level scope at REPL[7]:1
```

New error message:
```julia
julia> bounded(-0.1, 0.0, 1.0)
ERROR: ArgumentError: Value, -0.1, outside of specified bounds (0.0, 1.0).
Stacktrace:
 [1] bounded(::Float64, ::Float64, ::Float64) at /Users/willtebbutt/.julia/dev/ParameterHandling/src/parameters.jl:70
 [2] top-level scope at REPL[7]:1
```




Similarly:
```julia
julia> positive(-0.1)
ERROR: DomainError with -0.10000000000100001:
log will only return a complex result if called with a complex argument. Try log(Complex(x)).
Stacktrace:
 [1] throw_complex_domainerror(::Symbol, ::Float64) at ./math.jl:33
 [2] log(::Float64) at ./special/log.jl:285
 [3] Log at /Users/willtebbutt/.julia/packages/Bijectors/66bcj/src/bijectors/exp_log.jl:14 [inlined]
 [4] positive at /Users/willtebbutt/.julia/dev/ParameterHandling/src/parameters.jl:35 [inlined] (repeats 2 times)
 [5] top-level scope at REPL[8]:1
```
becomes
```julia
julia> positive(-0.1)
ERROR: ArgumentError: Provided val, -0.1, is not positive.
Stacktrace:
 [1] positive at /Users/willtebbutt/.julia/dev/ParameterHandling/src/parameters.jl:33 [inlined]
 [2] positive(::Float64) at /Users/willtebbutt/.julia/dev/ParameterHandling/src/parameters.jl:32
 [3] top-level scope at REPL[5]:1
```

Reply:
ah nice nice!